cloudinary.go

```diff
@@ -0,0 +1,390 @@
+	}
+
+	return result, nil
+}
+
+// AllFiles returns all the files from the given directory including all its subdirectories.
+func (r *Cloudinary) AllFiles(path string) ([]string, error) {
+	var result []string
+	assetTypes := []api.AssetType{api.Image, api.Video, api.File}
+	for _, assetType := range assetTypes {
+		nextCursor := ""
+		for {
+			response, err := r.instance.Admin.Assets(r.ctx, admin.AssetsParams{
+				Prefix:       validPath(path),
+				DeliveryType: "upload",
+				AssetType:    assetType,
+				MaxResults:   500,
+				NextCursor:   nextCursor,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, folder := range response.Assets {
+				result = append(result, folder.PublicID)
+			}
+
+			nextCursor = response.NextCursor
+			if nextCursor == "" {
+				break // Exit the loop when there is no next cursor
+			}
+		}
+	}
+	return result, nil
+}
+
+// Copy copies a file to a new location.
+func (r *Cloudinary) Copy(source, destination string) error {
+	result, err := r.instance.Upload.Upload(r.ctx, r.Url(source), uploader.UploadParams{
+		PublicID:     r.getPublicId(destination),
+		ResourceType: r.getResourceType(destination),
+	})
+	if err != nil {
+		return err
+	}
+	if result.Error.Message != "" {
+		return fmt.Errorf("copy file error: %#v", result.Error)
+	}
+	return nil
+}
+
+// Delete deletes a file.
+func (r *Cloudinary) Delete(file ...string) error {
+	for _, f := range file {
+		result, err := r.instance.Upload.Destroy(r.ctx, uploader.DestroyParams{
+			PublicID:     r.getPublicId(f),
+			ResourceType: r.getResourceType(f),
+			Invalidate:   api.Bool(true),
+		})
+		if err != nil {
+			return err
+		}
+		if result.Result != "ok" {
+			return fmt.Errorf("delete file error: %+v", result.Error)
+		}
+	}
+	return nil
+}
+
+// DeleteDirectory deletes a directory.
+func (r *Cloudinary) DeleteDirectory(directory string) error {
+	assetTypes := []api.AssetType{api.Image, api.Video, api.File}
+	for _, assetType := range assetTypes {
+		_, err := r.instance.Admin.DeleteAssetsByPrefix(r.ctx, admin.DeleteAssetsByPrefixParams{
+			Prefix:    []string{validPath(directory)},
+			AssetType: assetType,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	_, err := r.instance.Admin.DeleteFolder(r.ctx, admin.DeleteFolderParams{
+		Folder: directory,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Directories return all the directories within a given directory.
+func (r *Cloudinary) Directories(path string) ([]string, error) {
+	folders, err := r.instance.Admin.SubFolders(r.ctx, admin.SubFoldersParams{
+		Folder: validPath(path),
+	})
+	if err != nil {
+		return nil, err
+	}
+	var result []string
+	for _, folder := range folders.Folders {
+		result = append(result, folder.Path)
+	}
+	return result, nil
+}
+
+// Exists checks if a file exists in the Cloudinary storage.
+func (r *Cloudinary) Exists(file string) bool {
+	asset, err := r.getAsset(file)
+	if asset.Error.Message != "" {
+		return false
+	}
+	return err == nil
```

Comment:
S: The error needs to be judged first.

Reply:
I guess we don't need to judge it since we are not sending an error anyway. We just want to check if there is an error. If yes, we will return false; else, true.

Reply:
`assert` is a point type, it may be `nil` when `err != nil`, so we need to judge `err != nil` first.